File: Beecrowd/7-Grafos/1191-Recuperação_da_Árvore-5/1191.cs
using System;
using System.Collections.Generic;

class Program {
    static string FindPosfixo(string prefixo, string infixo) {
        int n = prefixo.Length;
        List<char> resultado = new List<char>();
        Stack<(int pre_idx, int in_start, int in_end)> pilha = new Stack<(int, int, int)>();
        
        pilha.Push((0, 0, n));
        
        while (pilha.Count > 0) {
            var (pre_idx, in_start, in_end) = pilha.Pop();
            
            if (in_start >= in_end) {
                continue;
            }
            
            char raiz = prefixo[pre_idx];
            int in_idx = infixo.IndexOf(raiz);
            
            pilha.Push((pre_idx + 1, in_start, in_idx));
            pilha.Push((pre_idx + (in_idx - in_start) + 1, in_idx + 1, in_end));
            
            resultado.Add(raiz);
        }
        
        resultado.Reverse();
        return new string(resultado.ToArray());
    }

    static void Main() {
        

        while(true) {
            try{
                string[] inputs = Console.ReadLine().Split();
                string prefixo = inputs[0];
                string infixo = inputs[1];
                
                string posfixo = FindPosfixo(prefixo, infixo);
                Console.WriteLine(posfixo);
            }
            catch{
                break;
            }
        }
    }
}

File: Beecrowd/7-Grafos/1191-Recuperação_da_Árvore-5/1191.py
def find_posfixo(prefixo, infixo):
    n = len(prefixo)
    posfixo = []
    pilha = [(0, 0, n)]
    
    resultado = []
    while pilha:
        pre_idx, in_start, in_end = pilha.pop()
        
        if in_start < in_end:
            raiz = prefixo[pre_idx]
            in_idx = infixo.index(raiz)
            
           
            pilha.append((pre_idx + 1, in_start, in_idx))
            pilha.append((pre_idx + (in_idx - in_start) + 1, in_idx + 1, in_end))
            
            resultado.append(raiz)
    
    return ''.join(resultado[::-1])


while True:
    try:
        prefixo, infixo = input().split()
        posfixo = find_posfixo(prefixo, infixo)
        print(posfixo)
    except:
        break

